provider/core/src/resource.rs

```diff
@@ -74,59 +73,132 @@ impl writeable::Writeable for ResourceCategory {
     }
 }
 
-/// A category, subcategory, and version, used for requesting data from a
-/// [`DataProvider`](crate::DataProvider).
-///
-/// The fields in a [`ResourceKey`] should generally be known at compile time.
+/// A compact hash of a [`ResourceKey`]. Useful for keys in maps.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Copy, Clone, Hash)]
+#[repr(transparent)]
+pub struct ResourceKeyHash([u8; 4]);
+
+impl ResourceKeyHash {
+    const fn compute_from_str(path: &str) -> Self {
+        Self(helpers::fxhash_32(path.as_bytes()).to_le_bytes())
+    }
+}
+
+/// The resource key used for loading data from an ICU4X data provider.
 ///
-/// Use [`resource_key!`] as a shortcut to create resource keys in code.
-#[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone)]
+/// A resource key is tightly coupled with the code that uses it to load data at runtime.
+/// Executables can be searched for ResourceKey instances to produce optimized data files.
+/// Therefore, users should not generally create ResourceKey instances; they should instead use
+/// the ones exported by a component.
+#[derive(PartialEq, Eq, Copy, Clone)]
+#[repr(C)]
 pub struct ResourceKey {
-    pub category: ResourceCategory,
-    pub sub_category: TinyStr16,
-    pub version: u16,
+    path: &'static str,
+    _tag0: [u8; 8],
+    hash: ResourceKeyHash,
+    _tag1: [u8; 2],
```

Comment:
I had the closing tag to make it less likely to have collisions (based on the assumption that one instance of a tag is more likely to occur than two tags separated by a specific number of bytes).

I'll need to check to make sure that `self.bytes[9..=12]` doesn't generate panicky code.  Since it's an array, theoretically it should figure out that the operation can't fail, but in general the index operator `[]` is panicky.

What's wrong with `repr(C)`?

Reply:
Slicing is actually done by a trait, so `self.bytes[9..=12]` isn't `const`.

Regarding `repr(C)` it just *seems* like it's the wrong hammer. We're not interacting with C code, we're just trying to put data in a specific order, which I think is clearer by using a single array if the array types match like they do here.

Reply:
> based on the assumption that one instance of a tag is more likely to occur than two tags separated by a specific number of bytes

I don't think that holds if the single tag has the same length as the two tags combined.